hal/src/common/dmac/transfer.rs

```diff
@@ -0,0 +1,857 @@
+    /// if a trigger request is already pending for the channel.
+    pub fn software_trigger(&mut self, dmac: &mut DmaController) {
+        self.chan.software_trigger(dmac.as_mut());
+    }
+    /// Blocking; Wait for the DMA transfer to complete and release all owned
+    /// resources
+    pub fn wait(self, dmac: &mut DmaController) -> (Src, Dst, Channel<Ready, ID>, P) {
+        let chan = self.chan.release(dmac.as_mut());
+
+        // Memory barrier to prevent the compiler/CPU from re-ordering read/write
+        // operations beyond this fence.
+        // (see https://docs.rust-embedded.org/embedonomicon/dma.html#compiler-misoptimizations)
+        atomic::fence(atomic::Ordering::Acquire); // ▼
+
+        (self.config.source, self.config.dest, chan, self.payload)
+    }
+
+    /// Non-blocking; Immediately stop the DMA transfer and release all owned
+    /// resources
+    pub fn stop(self, dmac: &mut DmaController) -> (Src, Dst, Channel<Ready, ID>, P) {
+        let chan = self.chan.stop(dmac.as_mut());
+
+        // Memory barrier to prevent the compiler/CPU from re-ordering read/write
+        // operations beyond this fence.
+        // (see https://docs.rust-embedded.org/embedonomicon/dma.html#compiler-misoptimizations)
+        atomic::fence(atomic::Ordering::Acquire); // ▼
+
+        (self.config.source, self.config.dest, chan, self.payload)
+    }
+}
+
+/// These methods are available to a `DmaTransfer` holding a `Busy` channel
+impl<B, P, Src, Dst, const ID: u8> DmaTransfer<B, P, Src, Dst, Busy, ID>
+where
+    B: 'static + DmaBeat,
+    Src: DmaBuffer<B>,
+    Dst: DmaBuffer<B>,
+{
+    /// Issue a software trigger request to the corresponding channel.
+    /// Note that is not guaranteed that the trigger request will register,
+    /// if a trigger request is already pending for the channel.
+    #[inline(always)]
+    pub fn software_trigger(&mut self, dmac: &mut DmaController) {
+        self.xfer.software_trigger(dmac);
+    }
+    /// Blocking; Wait for the DMA transfer to complete and release all owned
+    /// resources
+    #[inline(always)]
+    pub fn wait(self, dmac: &mut DmaController) -> (Src, Dst, Channel<Ready, ID>, P) {
+        self.xfer.wait(dmac)
+    }
+
+    /// Non-blocking; Immediately stop the DMA transfer and release all owned
+    /// resources
+    #[inline(always)]
+    pub fn stop(self, dmac: &mut DmaController) -> (Src, Dst, Channel<Ready, ID>, P) {
+        self.xfer.stop(dmac)
+    }
+}
+
+/// Useable beat sizes for DMA transfers
+#[derive(Clone, Copy)]
+pub enum BeatSize {
+    /// Byte = [`u8`](core::u8)
+    Byte = 0x00,
+    /// Half word = [`u16`](core::u16)
+    HalfWord = 0x01,
+    /// Word = [`u32`](core::u32)
+    Word = 0x02,
+}
+
+// Trait enabling taking `*mut` references to references to a single `T`,
+/// references to arrays of `T` or slices of `T`, where `T: DmaBeat`,
+/// as well as required transfer length
+pub trait DmaBuffer<T: DmaBeat> {
```

Comment:
Normally, I would use `Sealed` as a super trait for something like this. But ff you use the blanket implementation approach for types that implement `AsMut<[T]>`, then you won't be able to do that. I'd have to think a little bit more about how you can prevent people from adding their own implementations.